datasync/chaindatafetcher/kafka/consumer.go

```diff
@@ -101,20 +158,131 @@ func (c *Consumer) Cleanup(s sarama.ConsumerGroupSession) error {
 	return nil
 }
 
+// insertSegment inserts the given segment to the given buffer.
+// Assumption:
+//  1. it is guaranteed that the order of segments is correct.
+//  2. the inserted messages may be duplicated.
+//
+// We can consider the following cases.
+// case1. new segment with index 0 is inserted into newly created segment slice.
+// case2. new consecutive segment is inserted into the right position.
+// case3. duplicated segment is ignored.
+// case4. new sparse segment shouldn't be given, so return an error.
+func insertSegment(segment *Segment, buffer [][]*Segment) ([][]*Segment, error) {
+	for idx, segments := range buffer {
+		length := len(segments)
+		if length > 0 && segments[0].key == segment.key {
+			// the segment is already inserted to buffer.
+			if segment.index < uint64(length) {
+				logger.Warn("the message is duplicated", "segment", segment)
+				return buffer, nil
+			}
+
+			// the segment is inserted.
+			if segment.index == uint64(length) {
+				buffer[idx] = append(segments, segment)
+				return buffer, nil
+			}
+
+			// there is a missing segment which should not be occurred.
+			return buffer, errors.New(wrongSegmentErrorMsg)
+		}
+	}
+
+	if segment.index == 0 {
+		// create a segment slice and append it.
+		buffer = append(buffer, []*Segment{segment})
+	} else {
+		// the segment may be already handled.
+		logger.Warn("the message may be inserted already. drop the segment", "segment", segment)
```

Comment:
What's the difference between this line and line#177?

Reply:
For example, M1S0 M2S0 M1S1(read) M2S1 are in a partition and M1S1 is read, but not M2S1. M1 is reassembled and it has been processed. Then, the offset is updated to M2S0 as not consumed. In this case, M2S0 M1S1 M2S1 will be read in the order. However, M1 is already handled, so M1S1 should be ignored. That is the case of #197.

On the other hand, M1S0 M1S2 should not exist since M1S1 is missing. This case is line 177.

Reply:
There could be a case when a new message that doesn't have index 0 is inserted?

Reply:
> There could be a case when a new message that doesn't have index 0 is inserted?

@winnie-byun As far as I know, as the order of the messages is guaranteed, that should not happen.

Reply:
@winnie-byun @ehnuje So, if there is a message segment which doesn't have index 0, then the segment is handled as consumed message. At the above example, M1S0 |offset| M2S0 M1S1 M2S2 read from M2S0 and M1S1 in the order. But, M1S0 does not exist in the buffer, so the segment is ignored.